---
title: "BRN: RNA Seq Analysis"
output:
  html_document:
    fig_caption: yes
    toc: true
    toc_float:
      collapsed: false
      smooth_scroll: false
---
Based on the workshop performed in [Week 8 of BigBioinformatics R Programming and RNA-Seq Analysis](https://www.bigbioinformatics.org/r-and-rnaseq-analysis).

```{r, include=FALSE}
library(dplyr)
library(knitr)
```

## Scope
In this analysis, we want to know what effects knocking down the [EWSR1-FLI1 oncogene](https://en.wikipedia.org/wiki/EWS/FLI) will have on the transcriptome of Ewing sarcoma tumor cells.

WIP

## Data
We'll first start by reading in our counts stored as a `RangedSummarizedExperiment` object.
```{r}
rse <- readRDS("EwS.rds")
```

Checking our sample metadata, we have two sample conditions: A EF1 knockdown and our control.
```{r}
cd <- as.data.frame(
  SummarizedExperiment::colData(rse)
)
testit::assert("condition" %in% colnames(cd))
testit::assert(unique(cd$condition) == c("shCTR", "shEF1"))
```

### Condition
We'll now construct our `DESeqDataSet` separating our data by the column `condition`.

* Splits our dataset between between our controls and the EF1 knockdown samples.

```{r, message=FALSE}
deseq_dataset <- DESeq2::DESeqDataSet(rse, design = ~condition)
```

### QC
Next, we check whether our data showns any sign of batch effects.

This is accomplished by:

1. Log transforming the counts.

    * Use `rlog` to convert counts to regularized log scale for visualization.
    * PCA results based on features with highest variance.

      * On linear scale, high mean features dominate.
      * On log scale. small changes with low counts would dominate

        ```
        A gene with 2 reads == 200% more reads than a gene with 1 reads
        A gene with 101 reads == 1% more reads than a gene with 100 reads.
        ```

2. Performing PCA on the log transformed matrix of counts.

```{r, results='hide'}
pca <- DESeq2::rlog(deseq_dataset) %>% BiocGenerics::plotPCA()
pca
```

From this plot, we can see that the first PC (PC1) cleanly splits samples by our sample conditions.

* Because of this we know that the conditions explain the most variance.
* This gives us good confidence that any changes we observe in our analysis are due to our conditions rather than any technical variation.

## Analysis
We perform the differential expression analysis using [`DESeq2`](http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html).

And get our results showing relative change in expression of the knockdown compared to the control.

* `contrast` is a chr vector where 2nd and 3rd args are the numerator and denominator.

```{r, message=FALSE}
dds <- DESeq2::DESeq(deseq_dataset)
res <- DESeq2::results(dds, contrast = c("condition", "shEF1", "shCTR"))
```

### MA Plot
Next, we plot an MA plot.

* This shows the log2 fold changes attributable to a given variable over the mean of normalized counts for all the samples in the DESeqDataSet.^3^
* Answers question:
> How does fold change relate to how much gene is expressed.

First, we have to shrink the fold change of low count genes as shifts in variance can be extreme.
```{r, message=FALSE, fig.cap="Spurious Log Fold Changes at Low Count Genes", results='hide'}
plot_ma_origfc <- DESeq2::plotMA(res)
plot_ma_origfc
```

By shrinking the fold change: ^4^

* Creates prior distribution of lfcs from genes with high counts.
* Filters noise from low count genes.

```{r, message=FALSE}
# coef = 2 is the second item of resultsNames(dds)
res_shrunkfc <- DESeq2::lfcShrink(
  dds = dds,
  res = res,
  type = "normal",
  coef = 2
)
```

After lfc shrinkage, the points with spurious fold changes are moved toward the center.
```{r, message=FALSE, fig.cap="Adjusted Log Fold Change", warning=FALSE, results='hide'}
plot_ma_shrunkfc <- DESeq2::plotMA(res_shrunkfc)
plot_ma_shrunkfc
```

### Annotation
Next will annotate our data. Currently, it lists genes with their ENSMBL IDs which isn't helpful.
```
r$> head(res_shrunkfc)
log2 fold change (MAP): condition shEF1 vs shCTR
Wald test p-value: condition shEF1 vs shCTR
DataFrame with 6 rows and 6 columns
                     baseMean log2FoldChange     lfcSE      stat      pvalue        padj
                    <numeric>      <numeric> <numeric> <numeric>   <numeric>   <numeric>
ENSG00000000003.14 32363.1500     -0.6338445 0.1584261 -4.000851 0.000063115 0.000256938
ENSG00000000005.5      0.0000             NA        NA        NA          NA          NA
ENSG00000000419.12 48118.6436     -0.1795122 0.0834463 -2.151229 0.031458107 0.070859811
ENSG00000000457.13 15337.8714     -0.0609588 0.1335490 -0.456452 0.648065053 0.783818379
ENSG00000000460.16 23463.4429     -0.7325704 0.1939355 -3.777309 0.000158532 0.000604503
ENSG00000000938.12    13.5656     -0.4378801 0.4032424 -2.053576 0.040016781 0.086639736
```

We'll wrap our data wrangling operations in `annotate_deseq_res` where the following is performed:

1. Use the `AnnotationDbi` package to run a query on `EnsDB.Hsapiens.v86` and pull gene symbols mapping to Ensmbl IDs.
2. Pull each id to a new column `ENSID`
3. Clean the id removing trailing digits and set it a new `GENEID` column
4. Drop `ENSID`.
5. Perform an inner join only labeling genes by `GENEID`s that exist in both dataframes.

```{r}
# Annotate deseq gene IDs with symbols.
annotate_deseq_res <- function(resdf) {
  ens2sym <- AnnotationDbi::select(
    EnsDb.Hsapiens.v86::EnsDb.Hsapiens.v86,
    keys = AnnotationDbi::keys(EnsDb.Hsapiens.v86::EnsDb.Hsapiens.v86),
    columns = c("SYMBOL")
  )

  return(
    # nolint start
    resdf %>%
      tibble::rownames_to_column(var = "ENSID") %>%
      # Remove any trailing digits from IDs. ex. ########.##
      dplyr::mutate(GENEID = gsub(ENSID, pattern = "\\..+", replacement = "")) %>%
      dplyr::select(-ENSID) %>%
      dplyr::inner_join(y = ens2sym, by = "GENEID")
    # nolint end
  )
}
```

The output is saved as `resdf`.
```{r}
resdf <- as.data.frame(res_shrunkfc) %>% annotate_deseq_res()
```


## Results

### Volcano Plot
To get an overview of our differentially expressed genes, we'll plot a volcano plot.

* Volcano plots allow visualization of genes that have large fold changes (either upregulated or downregulated) that are significant.

```{r, message=FALSE, results='hide', fig.cap="Significant Differentially Expressed Genes"}
plot_volcano <- EnhancedVolcano::EnhancedVolcano(
  resdf,
  lab = resdf$SYMBOL,
  pCutoff = 1e-100,
  FCcutoff = 3,
  x = "log2FoldChange", y = "padj"
)
plot_volcano
```

### Top DE Genes
To get the top differentially expressed genes, we must first declare some cutoffs for valid genes.

* Has a **p-value of ±0.01** to be considered significant.
* Has a **log fold change of ±2** to be considered overexpressed or underexpressed, respectively.

```{r}
n_genes <- 10
p_thr <- 0.01
log_fold_change_thr <- 2
```

This code segment:

1. Filter our DE genes by p-value threshold of `r p_thr`
2. Assigns a new column `type` describing the expression level of the gene.
3. Sorts, in descending order, by `log2FoldChange`.
4. Adds a `rank` column.

```{r}
resdf_de_genes <- resdf %>%
  # Must be significant.
  dplyr::filter(padj < p_thr) %>%
  # Add column to describe expression level.
  dplyr::mutate(
    type = dplyr::case_when(
      log2FoldChange > log_fold_change_thr ~ "overexpressed",
      log2FoldChange < -log_fold_change_thr ~ "underexpressed",
      .default = "other"
    )
  ) %>%
  # Sort by log2 fold change.
  dplyr::arrange(desc(log2FoldChange)) %>%
  # And add a rank.
  dplyr::mutate(rank = row_number())
```

After wrangling and sorting the dataset, we can simply filter by `rank` to get the top `r n_genes` DE genes.
```{r}
num_de_genes <- nrow(resdf_de_genes)
resdf_top_de_genes <- resdf_de_genes %>%
  dplyr::filter(
    between(rank, 1, n_genes) |
      between(rank, num_de_genes - (n_genes - 1), num_de_genes)
  )
```

```{r echo=FALSE}
kable(
  resdf_top_de_genes,
  caption = sprintf("Top %i Differentially Expressed Genes", n_genes)
)
```
### Sources
1. https://www.bigbioinformatics.org/r-and-rnaseq-analysis
2. https://en.wikipedia.org/wiki/EWS/FLI
3. http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html
4. https://www.biostars.org/p/445113/
5. https://support.bioconductor.org/p/77461/#77466
6. https://training.galaxyproject.org/training-material/topics/transcriptomics/tutorials/rna-seq-viz-with-volcanoplot/tutorial.html
